##**An attempt to classify whether the news is Fake or Real by applying threshold on BaselineModel-1 and BaselineModel-2 prediction results and output the model predictions as either real or fake**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("D:\Downloads\Equal Real and fake claims - Sheet1.csv" , encoding='latin1') # Baseline model-1 & 2 prediction results

In [ ]:
df.head()

headline Type_of_claim  \
0  Utah woman delivers baby boy onboard flight to...          Real   
1  Portugal inaugurates 'world's longest' pedestr...          Real   
2  Stand-off continues with China on eastern fron...          Real   
3  PM Narendra Modi lays the foundation stone of ...          Real   
4            Communal riots in Delhi claims 42 lives          Real   

   Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  
0                      7        7          0      7         0        0  
1                      5        4          1      4         0        0  
2                     15        8          7      6         0        2  
3                     27       21          6     19         0        2  
4                     24       17          7      0         0       17

In [ ]:
#df.rename(columns = {'Related\n(Agree + discuss + Discuss)':'Related}, inplace = True)

In [ ]:
df.shape

(175, 8)

In [ ]:
df["related%"] = np.round((df["Related"]/df["Total_No._of_Articles"])*100)

In [ ]:
df.head()

headline Type_of_claim  \
0  Utah woman delivers baby boy onboard flight to...          Real   
1  Portugal inaugurates 'world's longest' pedestr...          Real   
2  Stand-off continues with China on eastern fron...          Real   
3  PM Narendra Modi lays the foundation stone of ...          Real   
4            Communal riots in Delhi claims 42 lives          Real   

   Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
0                      7        7          0      7         0        0   
1                      5        4          1      4         0        0   
2                     15        8          7      6         0        2   
3                     27       21          6     19         0        2   
4                     24       17          7      0         0       17   

   related%  
0     100.0  
1      80.0  
2      53.0  
3      78.0  
4      71.0

In [ ]:
# applying the threshold of 51% on 2-class prediction (BaselineModel-1), only the claims that are having related articles > 15% are passed to the BaselineModel-2
df_M2 = df[df["related%"] > 51.0] 

In [ ]:
df_M2["Type_of_claim"].value_counts()

Real      81
Unreal    46
Name: Type_of_claim, dtype: int64

In [ ]:
df_M1_out = df[df["related%"] <= 51.0]
df_M1_out.shape
df_M1_out["Type_of_claim"].value_counts()

Unreal    28
Real      20
Name: Type_of_claim, dtype: int64

In [ ]:
df_M2.head()

headline Type_of_claim  \
0  Utah woman delivers baby boy onboard flight to...          Real   
1  Portugal inaugurates 'world's longest' pedestr...          Real   
2  Stand-off continues with China on eastern fron...          Real   
3  PM Narendra Modi lays the foundation stone of ...          Real   
4            Communal riots in Delhi claims 42 lives          Real   

   Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
0                      7        7          0      7         0        0   
1                      5        4          1      4         0        0   
2                     15        8          7      6         0        2   
3                     27       21          6     19         0        2   
4                     24       17          7      0         0       17   

   related%  
0     100.0  
1      80.0  
2      53.0  
3      78.0  
4      71.0

In [ ]:
# if count of agree is less than sum of discuss and disagree, then we output the claim as Fake news
df[df["Agree"] < (df["Discuss"] + df["Disagree"])]

headline Type_of_claim  \
4              Communal riots in Delhi claims 42 lives          Real   
8    Kerala CM Pinarayi Vijayan alleges largescale ...          Real   
20   If 30% Muslims unite in India, 4 Pakistans wil...          Real   
34   "Tim Allen quote Trump’s wall costs less than ...        Unreal   
35   "Democrats Vote To Enhance Med Care for Illega...        Unreal   
39   "Trump’s grandfather was a pimp and tax evader...        Unreal   
40   Bill Gates has created or financed the creatio...        Unreal   
45   Kashmiri Muslim woman called ‘terrorist’ by he...        Unreal   
48    Hindus are behind Tiktok videos urging Muslim...        Unreal   
51   china accepts 30 soldiers have been killed dur...        Unreal   
52   omar abdullah said If Article 370 goes, India-...        Unreal   
55   Chinese Intelligence Officer Revealed that Cor...        Unreal   
59   Cyclone Tauktae: 56 trains cancelled as storm ...          Real   
63   Assam Minister orders probe into death of 18 e...          Real   
67   Onset of southwest monsoon over Kerala likely ...          Real   
71   Militant who killed Assam Rifles jawan in 2017...          Real   
76              King of Bahrain King's Robot Bodyguard        Unreal   
78   china accepts 30 soldiers have been killed dur...        Unreal   
79   omar abdullah said If Article 370 goes, India-...        Unreal   
81   Chinese Intelligence Officer Revealed that Cor...        Unreal   
91   Statements made in Morgan Stanley Research pra...        Unreal   
92   National Geographic magazine cover on recent f...        Unreal   
99   Former Pakistani Diplomat admits to 300 casual...        Unreal   
100  BJP leaders pleading Rajnath Singh to revoke t...        Unreal   
101  Newspaper clipping with a report of 40 student...        Unreal   
104  Supreme Court has approved teachers to put the...        Unreal   
107              Birds Dying Due To Jio’s 5G Testing        Unreal   
115  Taj Hotel is sending gift cards for a seven-da...        Unreal   
121  1300-year-old bicycle sculpture is carved on t...        Unreal   
126  148 samples from karnataka found infected with...          Real   
133  Man place bids on women as Indian youtube chan...          Real   
137  UP couple sell 3 month old baby boy fro 1.5 La...          Real   
138  73 Million year old talkative dinosaur species...          Real   
153  Mumbai man orders mouthwash from Amazon, gets ...          Real   
159  Several leaders’ phones tapped during Fadnavis...          Real   
170  China says price hike in anti-Covid medical su...          Real   

     Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
4                       24       17          7      0         0       17   
8                        9        8          1      0         0        8   
20                      10        6          4      0         2        4   
34                       3        2          1      0         1        1   
35                       2        1          1      0         0        1   
39                       7        5          2      0         3        2   
40                       9        8          1      3         4        1   
45                       6        6          0      2         0        4   
48                       8        2          6      0         2        0   
51                      18       14          4      4         1        9   
52                       8        4          4      0         0        4   
55                      17        8          9      3         3        2   
59                       9        4          5      0         0        4   
63                      12       11          1      1         0       10   
67                      22       22          0      0         0       22   
71                       9        5          4      0         0        5   
76                       4        3          1      0         1        2   
78           

In [ ]:
# if count of agree is less than sum of discuss and disagree, and count of discuss is less than the count of disagree, then we output the claim as Fake news


pred = []
for i in range(len(df_M2)):
    if (df_M2["Agree"].iloc[i] < (df_M2["Discuss"].iloc[i] + df_M2["Disagree"].iloc[i])):
        if df_M2["Discuss"].iloc[i] <= df_M2["Disagree"].iloc[i]:
            pred.append("Unreal")
        else:
            pred.append("Real")
    else:
        pred.append("Real")
df_M2["pred"] = pred
print(len(pred))

        
   

127


C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # Remove the CWD from sys.path while we load stuff.


In [ ]:
df_M2

headline Type_of_claim  \
0    Utah woman delivers baby boy onboard flight to...          Real   
1    Portugal inaugurates 'world's longest' pedestr...          Real   
2    Stand-off continues with China on eastern fron...          Real   
3    PM Narendra Modi lays the foundation stone of ...          Real   
4              Communal riots in Delhi claims 42 lives          Real   
..                                                 ...           ...   
166  Mamata Banerjee's brother Ashim Bandhopadhay p...          Real   
167  Centre to ramp up vaccine supply; Health Minis...          Real   
171  Centre distributing vaccines rationally, Benga...          Real   
172  One body recovered from Tapovan tunnel in Utta...          Real   
173  Odisha CM Naveen Patnaik releases Rs 920 crore...          Real   

     Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
0                        7        7          0      7         0        0   
1                        5        4          1      4         0        0   
2                       15        8          7      6         0        2   
3                       27       21          6     19         0        2   
4                       24       17          7      0         0       17   
..                     ...      ...        ...    ...       ...      ...   
166                     21       15          6     10         0        5   
167                     23       19          4     17         0        2   
171                      2        2          0      2         0        0   
172                     11       11          0     11         0        0   
173                      5        3          2      3         0        0   

     related%  pred  
0       100.0  Real  
1        80.0  Real  
2        53.0  Real  
3        78.0  Real  
4        71.0  Real  
..        ...   ...  
166      71.0  Real  
167      83.0  Real  
171     100.0  Real  
172     100.0  Real  
173      60.0  Real  

[127 rows x 10 columns]

In [ ]:
df_M2['Type_of_claim'].value_counts()

Real      81
Unreal    46
Name: Type_of_claim, dtype: int64

In [ ]:
df_M2['pred'].value_counts()

Real      120
Unreal      7
Name: pred, dtype: int64

In [ ]:
nf = df_M2.copy()

In [ ]:
pred = []
for i in range(len(nf)):
    if (nf["Agree"].iloc[i] < (nf["Discuss"].iloc[i] + nf["Disagree"].iloc[i])):
        pred.append("Unreal")
    elif nf["Discuss"].iloc[i] <= nf["Disagree"].iloc[i]:
        pred.append("Unreal")
    else:
        pred.append("Real")
nf["pred"] = pred
print(len(pred))


127


In [ ]:
nf['Type_of_claim'].value_counts()

Real      81
Unreal    46
Name: Type_of_claim, dtype: int64

In [ ]:
nf['pred'].value_counts()

Unreal    82
Real      45
Name: pred, dtype: int64

In [ ]:
nf["flag"] = nf["Type_of_claim"] == nf["pred"]
nf.head()

headline Type_of_claim  \
0  Utah woman delivers baby boy onboard flight to...          Real   
1  Portugal inaugurates 'world's longest' pedestr...          Real   
2  Stand-off continues with China on eastern fron...          Real   
3  PM Narendra Modi lays the foundation stone of ...          Real   
4            Communal riots in Delhi claims 42 lives          Real   

   Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
0                      7        7          0      7         0        0   
1                      5        4          1      4         0        0   
2                     15        8          7      6         0        2   
3                     27       21          6     19         0        2   
4                     24       17          7      0         0       17   

   related%    pred   flag  
0     100.0  Unreal  False  
1      80.0  Unreal  False  
2      53.0    Real   True  
3      78.0    Real   True  
4      71.0  Unreal  False

In [ ]:
nf[nf["Type_of_claim"] =="Unreal"]["flag"].value_counts()

True     27
False    19
Name: flag, dtype: int64

In [ ]:
nf[nf["Type_of_claim"] =="Real"]["flag"].value_counts()

False    55
True     26
Name: flag, dtype: int64

In [ ]:
df_M1_out['pred'] = "Unreal"

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
ff = pd.concat([df_M1_out, nf])

In [ ]:
ff.head()

headline Type_of_claim  \
14  70 women earned nods, the first Muslim actor n...          Real   
24  Democracy’s “watchdog”: SC refuses to muzzle m...          Real   
35  "Democrats Vote To Enhance Med Care for Illega...        Unreal   
36  "Ilhan Omar Holding Secret Fundraisers With Is...        Unreal   
38  "Nancy Pelosi diverting Social Security money ...        Unreal   

    Total_No._of_Articles  Related  Unrelated  Agree  Disagree  Discuss  \
14                      5        2          3      2         0        0   
24                      3        1          2      1         0        0   
35                      2        1          1      0         0        1   
36                      3        0          3      0         0        0   
38                      5        2          3      1         0        1   

    related%    pred flag  
14      40.0  Unreal  NaN  
24      33.0  Unreal  NaN  
35      50.0  Unreal  NaN  
36       0.0  Unreal  NaN  
38      40.0  Unreal  NaN

In [ ]:
ff.shape

(175, 11)

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from nltk.corpus import stopwords
from sklearn.metrics import classification_report, confusion_matrix


In [ ]:
y = ff["Type_of_claim"]
y_pred = ff["pred"]

In [ ]:
print(classification_report(y,y_pred))
print(confusion_matrix(y,y_pred))

              precision    recall  f1-score   support

        Real       0.58      0.26      0.36       101
      Unreal       0.42      0.74      0.54        74

    accuracy                           0.46       175
   macro avg       0.50      0.50      0.45       175
weighted avg       0.51      0.46      0.43       175

[[26 75]
 [19 55]]


In [ ]:
81/175

0.46285714285714286

###**Conclusion :  As a result of thresholding the prediction output, we improved recall value for Fake news to 0.74 by making the model prediction biased towards detecting Fake News more accurately compared to real news**